### PR TITLE
Implement FTP file handler to support OTA

### PR DESCRIPTION
**Pull request by azeemshatp**
What does this PR do?
- Implement new functions to support file download from FTP server using FTPGETTOFS, delete and read the downloaded files from the SIM800 flash.
- Removed the buffer resp_buf[] from the SIM800 library to the external file, to follow encapsulation. 
- Adjusted the time delay in the function enable_bearer().

Where should the reviewer start?
SIM800-AT.cpp

### CODE REVIEW

File: src/SIM800-AT-Zephyr.cpp [hunk 1]
```diff
@@ -937,6 +938,85 @@ bool GPRS::get_location(float *latitude, float *longitude)
     return false;
 }
 
+int GPRS::delete_file(const char *file_name)
+{
+    // Delete the existing file in the SIM800 memory
+    char cmd[64];
+    snprintf(cmd, sizeof(cmd), "AT+FSDEL=C:\\User\\FTP\\%s", file_name);
+    printf("%s\r\n", cmd);
```
Comment (JordanSilverman) [inline src/SIM800-AT-Zephyr.cpp:946]: Don't use printf statements in this library

File: src/SIM800-AT-Zephyr.cpp [hunk 2]
```diff
@@ -937,6 +938,85 @@ bool GPRS::get_location(float *latitude, float *longitude)
...
+    k_sleep(K_SECONDS(1));
+    if (ack_received == false) {
+        return MODEM_RESPONSE_ERROR;
+    }
+
+    snprintf(cmd, sizeof(cmd), "AT+FTPGETTOFS=0,\"%s\"", file_name);
+    send_cmd(cmd, 90, NULL); // More than 1 minute for response?
+    k_sleep(K_SECONDS(1));
+
+    // If the response is as expected
```
Comment (JordanSilverman) [inline src/SIM800-AT-Zephyr.cpp:1005]: Doesn't look like you're actually checking for a response here

File: src/SIM800-AT-Zephyr.cpp [hunk 3]
```diff
@@ -937,6 +938,85 @@ bool GPRS::get_location(float *latitude, float *longitude)
...
+    return MODEM_RESPONSE_OK;
+}
+
+int GPRS::read_file(const char *file_name, int length, int offset)
+{
+    char cmd[64];
+    snprintf(cmd, sizeof(cmd), "AT+FSREAD=C:\\User\\FTP\\%s,1,%d,%d", file_name, length, offset);
+
+    send_cmd(cmd, 2, NULL);
```
Comment (JordanSilverman) [inline src/SIM800-AT-Zephyr.cpp:1014]: Again not checking for a response? What if the file doesn't exist?

File: src/SIM800-AT-Zephyr.cpp [hunk 4]
```diff
@@ -937,6 +938,85 @@ bool GPRS::get_location(float *latitude, float *longitude)
...
+    k_sleep(K_SECONDS(1));
+
+    // If the response is as expected
+    return MODEM_RESPONSE_OK;
+}
+
+int GPRS::read_file(const char *file_name, int length, int offset)
+{
+    char cmd[64];
+    snprintf(cmd, sizeof(cmd), "AT+FSREAD=C:\\User\\FTP\\%s,1,%d,%d", file_name, length, offset);
```
Comment (JordanSilverman) [inline src/SIM800-AT-Zephyr.cpp:1012]: This seems to be a generic read file function so a file might not be in C:\User\FTP\. Maybe you want to rename the function read_ftp_file

**Review by JordanSilverman — COMMENTED** [review]
A couple of comments / questions but in general it looks good, nice one.

File: src/SIM800-AT-Zephyr.cpp (hunk 1 above)
Comment (azeemshatp) [reply src/SIM800-AT-Zephyr.cpp:946]: Ok, I will remove it.
But how about this one, which I kept for signalling an invalid string length? 
https://github.com/lastmileICT/SIM800-AT/blob/OTA_Comm/src/SIM800-AT-Zephyr.cpp#L126

File: src/SIM800-AT-Zephyr.cpp (hunk 2 above)
Comment (azeemshatp) [reply src/SIM800-AT-Zephyr.cpp:1005]: There wont be any response to this command immediately, and hence, this is checked externally in the GSM state machine. I will describe this in the header file

File: src/SIM800-AT-Zephyr.cpp (hunk 3 above)
Comment (azeemshatp) [reply src/SIM800-AT-Zephyr.cpp:1014]: Similar reason as above. The calling party will use this function just to send the command and then wait for the buffer to get populated.

File: src/SIM800-AT-Zephyr.cpp (hunk 4 above)
Comment (azeemshatp) [reply src/SIM800-AT-Zephyr.cpp:1012]: That's correct. I will change it.

File: src/SIM800-AT-Zephyr.cpp (hunk 3 above)
Comment (azeemshatp) [reply src/SIM800-AT-Zephyr.cpp:1014]: We will have to add the response check in the calling function instead, and we also have plans to modify this to read larger chunks of bytes. So I would leave this now and improve in the next PR. Is that fine?

**Comment by martinjaeger** [comment]
I don't fully understand how that commit improves data encapsulation. Can you please explain?

**Comment by azeemshatp** [comment]
> I don't fully understand how that commit improves data encapsulation. Can you please explain?
This doesn't address the encapsulation now, so I am going to delete that part from the commit. In order not to delay the OTA module, I am deferring this to another PR. 
So the idea for the next PR would be to have the ISR update a private buffer as and when bytes are coming in. Define another function that would read this buffer and wait until an acknowledgement string is received, and so on. 
This function may have an external buffer pointer as the argument, and will be called from an external thread. Please let me know if you have more inputs on this.

**Review by martinjaeger — APPROVED** [review]
LGTM.

**Review by JordanSilverman — APPROVED** [review]
👍

**Pull request merged**